Mathlib/GroupTheory/Rank.lean

```diff
@@ -0,0 +1,84 @@
+/-
+Copyright (c) 2025 Thomas Browning. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Thomas Browning
+-/
+import Mathlib.GroupTheory.Finiteness
+import Mathlib.SetTheory.Cardinal.Finite
+
+/-!
+# Rank of a group
+
+This file defines the rank of a group, namely the minimum size of a generating set.
+-/
+
+open Function Group
+
+variable {G H : Type*} [Group G] [Group H]
+
+namespace Group
+
+variable (G) in
+/-- The minimum number of generators of a group. -/
+@[to_additive "The minimum number of generators of an additive group."]
+noncomputable def rank [h : FG G] : ℕ := @Nat.find _ (Classical.decPred _) (fg_iff'.mp h)
```

Comment:
Suggestion for the future: should we return a junk value (probably, zero) for non-finitely generated groups? Does it make sense to add `ENat` or `Cardinal`-valued versions?

Thanks!
bors d+